contracts/ExitFormat.sol

```diff
@@ -23,14 +18,35 @@ library ExitFormat {
 
     // A SingleAssetExit specifies
     // * an asset address (0 implies the native asset of the chain: on mainnet, this is ETH)
-    // * custom metadata (optional field, can be zero bytes). This might specify how to transfer this particular asset (e.g. target an "ERC20.transfer"' method)
+    // * custom tokenMetadata
+    //   containing
+    //   - AssetType enum value
+    //   - metadata for that token
     // * an allocations array
     struct SingleAssetExit {
         address asset;
-        bytes metadata;
+        TokenMetadata tokenMetadata;
         Allocation[] allocations;
     }
 
+    // TokenMetadata allows for different token standards
+    // that require additional data than just a token contract address
+    // * assetType specifies one of the supported asset types
+    // * metadata is a differently encoded metadata depending on the token type.
+    //   This is untyped to allow for extensions in future as different token standards emerge
+    struct TokenMetadata {
+        AssetType assetType;
+        bytes metadata;
+    }
+
+    // Enum of different (non-native) token types the SingleAssetExit can contain
+    enum AssetType {Null, ERC20, ERC721, ERC1155}
+
+    // Metadata structure for ERC721 and ERC1155 exits
+    struct TokenIdExitMetadata {
+        uint256 tokenId;
+    }
+
     // allocations is an ordered array of Allocation.
     // The ordering is important, and may express e.g. a priority order for the exit
     // (which would make a material difference to the final state in the case of running out of gas or funds)
```

Comment:
Outside the scope of this PR, but somewhat higher priority because of it: we should specify how `executeSingleAssetExit` and `executeExit` behave when an asset transfer fails. I believe currently this will revert the entire transaction, which opens up griefing vectors where parties slip in "bad" assets and ruin things for everyone.

Reply:
Yeah this is an interesting one. Asset transfers could previously only fail due to:
- an invalid asset address
- insufficient funds

They can now also fail due to:
- Invalid TokenId
- Receiver not implementing `IERC1155Receiver` (if a contract)
- Receiver contract actively rejecting receiving ERC1155 which can be conditional on the payload

The last point is particularly interesting and could allow some attacks if not handled properly. For example a malicious participant in a channel could use a recipient address that rejects transfers unless the owner enables them. If an outcome is finalized that transfers to that recipient then the malicious participant can block all withdrawals and extort the other participant!

Reply:
I guess could verify the receiver address and ensure it is an EOA. That way the safe transfer cannot fail for 1155 and 721

Reply:
I think the case of a "bad" ERC20 implementation could be added to the first of your lists. But yes, I think there are now some more failure modes. 

Requiring beneficiaries to be EOAs is an interesting approach, but:
1. Is it even possible to do this reliably? Is an EOA distinguishable from a `CREATE2` address with no deployed code (yet)?
2. Even if it is possible, this seems to restrict the use cases of the exit format. One such use case would be "pay out directly to Uniswap" (for example). 

Another way to mitigate the risk is to have (out of band) whitelists of approved / vetted token contracts, generated by some authority or community.

I think overall we don't need to solve this problem right now, but we should at least document how the code here will behave. For example a comment such as "if one token transfer fails, the entire exit execution will revert".

Reply:
Happy for this to tackled in a follow-up PR.

Reply:
> Is it even possible to do this reliably?

What I was suggesting is not that the exit code checks but rather that any participant engaging with the channel checks before they participate. I don't think you can check that an address is an EOA from on-chain, only that an address is an existing contract. You can see how the check is done for `safeTransferFrom` [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/b0cf6fbb7a70f31527f36579ad644e1cf12fdf4e/contracts/utils/Address.sol#L26)

From off-chain though an observer could see if the chain history contains transactions originating from an account and be convinced that it is an EOA.